File: Diameter of Binary Tree - Leetcode 543.py
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        largest_diameter = [0]

        def height(root):
            if root is None:
                return 0

            left_height = height(root.left)
            right_height = height(root.right)
            diameter = left_height + right_height

            largest_diameter[0] = max(largest_diameter[0], diameter)
            
            return 1 + max(left_height, right_height)

        height(root)
        return largest_diameter[0]

# Time Complexity: O(n)
# Space Complexity: O(h) { here "h" is the height of the tree }

File: Diameter of Binary Tree - Leetcode 543/Diameter of Binary Tree - Leetcode 543.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val);
 *     this.left = (left===undefined ? null : left);
 *     this.right = (right===undefined ? null : right);
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    let largestDiameter = 0;

    function height(node) {
        if (!node) return 0;
        
        const leftHeight = height(node.left);
        const rightHeight = height(node.right);
        const diameter = leftHeight + rightHeight;
        
        largestDiameter = Math.max(largestDiameter, diameter);
        
        return 1 + Math.max(leftHeight, rightHeight);
    }

    height(root);
    return largestDiameter;
};

// Time Complexity: O(n)
// Space Complexity: O(h) // "h" is the height of the tree
